SG14/inplace_function.h

```diff
@@ -29,340 +29,308 @@
-	{
-		throw std::bad_function_call();
-	}
+  vtable_ptr_t vtable_ptr_;
+  mutable storage_t storage_;
 
-	void set(std::nullptr_t)
-	{
-		m_ManagerFctPtr = nullptr;
-		m_InvokeFctPtr = &DefaultFunction;
-	}
-
-	// For function pointers
-	void set(CompatibleFunctionPointer ptr)
-	{
-		// this is dodgy, and - according to standard - undefined behaviour. But it works
-		// see: http://stackoverflow.com/questions/559581/casting-a-function-pointer-to-another-type
-		m_ManagerFctPtr = nullptr;
-		m_InvokeFctPtr = reinterpret_cast<InvokeFctPtrType>(ptr);
-	}
-
-	// Set - for functors
-	// enable_if makes sure this is excluded for function references and pointers.
-	template<typename FunctorArgT>
-	typename std::enable_if<!std::is_pointer<FunctorArgT>::value && !std::is_function<FunctorArgT>::value>::type
-	set(const FunctorArgT& ftor)
-	{
-		using FunctorT = typename std::remove_reference<FunctorArgT>::type;
-		static_assert(sizeof(FunctorT) <= Capacity, "Functor too big to fit in the buffer");
-
-		// copy functor into the mem buffer
-		FunctorT* buffer = reinterpret_cast<FunctorT*>(&m_Data);
-		new (buffer) FunctorT(ftor);
-
-		// generate destructor, copy-constructor and move-constructor
-		m_ManagerFctPtr = &manage<FunctorT>;
-
-		// generate entry call
-		m_InvokeFctPtr = &invoke<FunctorT>;
-	}
-
-	// Set - for functors
-	// enable_if makes sure this is excluded for function references and pointers.
-	template<typename FunctorArgT>
-	typename std::enable_if<!std::is_pointer<FunctorArgT>::value && !std::is_function<FunctorArgT>::value>::type
-	set(FunctorArgT&& ftor)
-	{
-		using FunctorT = typename std::remove_reference<FunctorArgT>::type;
-		static_assert(sizeof(FunctorT) <= Capacity, "Functor too big to fit in the buffer");
-
-		// copy functor into the mem buffer
-		FunctorT* buffer = reinterpret_cast<FunctorT*>(&m_Data);
-		new (buffer) FunctorT(std::move(ftor));
-
-		// generate destructor, copy-constructor and move-constructor
-		m_ManagerFctPtr = &manage<FunctorT>;
-
-		// generate entry call
-		m_InvokeFctPtr = &invoke<FunctorT>;
-	}
-
-	template <typename FunctorT>
-	static RetT invoke(ArgsT... args, const void* dataPtr)
-	{
-		FunctorT* functor = (FunctorT*)const_cast<void*>(dataPtr);
-		return (*functor)(std::forward<ArgsT>(args)...);
-	}
-
-	template <typename FunctorT>
-	static void manage(void* dataPtr, const void* fromPtr, Operation op)
-	{
-		FunctorT* thisFunctor = reinterpret_cast<FunctorT*>(dataPtr);
-		switch (op)
-		{
-		case Operation::Destroy:
-			thisFunctor->~FunctorT();
-			break;
-		case Operation::Copy:
-			{
-				const FunctorT* source = (const FunctorT*)const_cast<void*>(fromPtr);
-				new (thisFunctor) FunctorT(*source);
-				break;
-			}
-		case Operation::Move:
-			{
-				FunctorT* source = (FunctorT*)fromPtr;
-				new (thisFunctor) FunctorT(std::move(*source));
-				break;
-			}
-		}
-	}
+  inplace_function(vtable_ptr_t vtable_ptr, storage_t&& storage) noexcept :
+    vtable_ptr_{vtable_ptr}, storage_{storage}
```

Comment:
I don't understand why `storage` is an rvalue reference here.
I am unsure of whether `storage_{storage}` is guaranteed to do something-akin-to-memcpy or whether it's allowed to be a compiler error. Brace-initializing `aligned_storage_t` is an idiom I haven't seen before.

Reply:
Fixed with direct inplace construction and copy, move awareness.